### PR TITLE
PBL-32307 Font APIs

### PR DESCRIPTION
**Pull request by HBehrens**
Introduces 
`graphics_context_set_text_color()`, `graphics_draw_text()`, `graphics_text_layout_get_content_size()`, `fonts_get_system_font()` + system font keys, `fonts_load_custom_font_with_data()`, `fonts_load_custom_font()` and explains the basic usage in the new `examples/text/index.html` example

### CODE REVIEW

**Comment by HBehrens** [comment]
Looks like we're running into a similar issue to https://github.com/kripken/emscripten/issues/3945 - when uglifying `transpiled.js` any string passed to applib becomes `null`. I will investigate this further tomorrow, @franc0is

**Comment by HBehrens** [comment]
Quick update: It seems that my local emsdk is older than the date of this fix. Unfortunately, I get different errors when transpiling applib with `emscripten-tag-1.35.21-64bit`, `emscripten-tag-1.35.22-64bit`, or `emscripten-master-64bit`. Will look for a workaround by manually reimplementing `stringToC()` tomorrow before looking for an actual solution.

**Comment by HBehrens** [comment]
As I couldn't get the newer version of Emscripten to work with our setup, I applied the aforementioned patch as part of D4541. See comments over there, @franc0is .

File: test/manual/gbitmap.js [hunk 1]
```diff
@@ -200,11 +207,19 @@ describe('GBitmap', function() {
         delete symbols.Resources.defaultProxy;
```
Comment (franc0is) [inline test/manual/gbitmap.js]: This is a misnomer. Instead, should be "returns the url directly when no proxy is set"

File: test/manual/gfont.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    before(function() {
+      fonts_get_system_font = rocky.fonts_get_system_font;
+    });
+
+    it('returns null for some cases', function() {
+      expect(fonts_get_system_font(undefined)).to.be.null;
+      expect(fonts_get_system_font(null)).to.be.null;
+    });
+
+    xit('returns callback font for unknown keys', function() {
```
Comment (franc0is) [inline test/manual/gfont.js]: xit?

Comment (franc0is) [inline test/manual/gfont.js]: xit?

File: test/manual/gfont.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        sandbox.mock(rocky.module)
+          .expects('ccall').never();
+        var cPtr = font.captureCPointer();
+        expect(cPtr).to.equal(0);
+      });
+
+      it('captureCPointer', function() {
+        sandbox.mock(rocky.module.Runtime)
+          .expects('addFunction').twice()
+          .onCall(0).returns(123)
```
Comment (franc0is) [inline test/manual/gfont.js]: Maybe comments  as to what those calls should map to?

File: test/manual/gfont.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        expect(ccall.getCall(1).args).to.eql(
+          ['fonts_load_custom_font', 'number', ['number'], [789]]);
+      });
+
+      it('releaseCPointer', function() {
+        var ccall = sandbox.mock(rocky.module)
+          .expects('ccall').twice();
+        sandbox.mock(rocky.module.Runtime)
+          .expects('removeFunction').twice()
+          .onCall(0).returns(456)
```
Comment (franc0is) [inline test/manual/gfont.js]: returns? Or should it be expected an argument rather?

File: test/manual/gfont.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        // sinon's mocks don't seem to work with multiple expectations on the same mock
+        // hence this check afterwards
+        expect(ccall.getCall(0).args).to.eql(
+          ['fonts_unload_custom_font', 'void', ['number'], [321]]);
+        expect(ccall.getCall(1).args).to.eql(
+          ['emx_resources_remove_custom', 'void', ['number'], [789]]);
+      });
+    });
+
+    describe('fonts_load_custom_font', function() {
```
Comment (franc0is) [inline test/manual/gfont.js]: A test for the failure path (butkus returns a 400?) would be a good idea to have as well.

File: src/symbols-manual.js [hunk 6]
```diff
@@ -287,6 +312,100 @@ Rocky.addManualSymbols = function(obj) {
...
+    var sysFont = obj.module.ccall('fonts_get_system_font', 'number',
+                                  ['string'], [name]);
+    if (!sysFont) {
+      return null;
+    }
+
+    return {
+      status: obj.Resources.status.loaded,
+      captureCPointer: function() {
+        return sysFont;
```
Comment (franc0is) [inline src/symbols-manual.js]: Oh, the magic of garbage collected memory.

**Comment by franc0is** [comment]
:+1:

File: test/manual/gfont.js (hunk 2 above)
Comment (HBehrens) [inline test/manual/gfont.js]: That's Mocha's way of disable a test (`xit()`) or a whole set of tests (`xdescribe()`).

File: test/manual/gfont.js (hunk 4 above)
Comment (HBehrens) [inline test/manual/gfont.js]: :check

File: test/manual/gfont.js (hunk 5 above)
Comment (HBehrens) [inline test/manual/gfont.js]: Both, `GBitmap` and `GFont` are treated as `resourceObject` in `resourceObjectSetStatusAndCallEvents()` – it's tested for `GBitmap`. I will extract into an isolated test when I do the 3rd resource-based object (PDC or ImageSequence).

File: src/symbols-manual.js (hunk 6 above)
Comment (HBehrens) [inline src/symbols-manual.js]: I'd rather say: JS'es magic of anonymous functions and "closures" :)

File: test/manual/gbitmap.js (hunk 1 above)
Comment (HBehrens) [inline test/manual/gbitmap.js]: this test covers various combination of proxy and no proxy.

**Pull request merged**